### PR TITLE
fix(p2p): DNS-validate discovered DataHub URLs; probe all endpoints for truthful health

### PR DESCRIPTION
**Pull request by galt-tr**
## Problem

`https://arcade-v2-us-1.bsvblockchain.tech/health` reports `{url: "http://asset:8090/api/v1", source: "discovered", healthy: true}` — a peer's cluster-internal service name, unreachable from arcade. Two independent gaps:

1. **Discovery validation was syntactic-only** (`services/p2p_client/node_status.go`): the private-address check applied to IP *literals* only, so a bare hostname like `asset` passed without any check that it resolves or where it resolves to — a gap the old test suite explicitly documented as intended behavior. The URL then persisted forever in the shared registry (no TTL/eviction) and re-entered every pod's endpoint set on every 30s refresh.
2. **Health state never updated outside the propagation pod**: the recovery probe only targeted *already-unhealthy* endpoints, and only broadcast outcomes recorded failures. The api-server pod (which serves `/health`) never broadcasts, so its per-pod circuit-breaker map reported every endpoint `healthy: true` forever.

## Fix

Mirrors merkle-service #181's announcement-validation pattern:

- **New `ssrfguard` package** (ported from merkle-service; fixed the CGNAT 100.64/10 doc-vs-code gap during the port): scheme allowlist, userinfo rejection, cloud-metadata deny list, blocked-IP predicate, and DNS resolution of non-literal hostnames with **every** resolved IP checked. **Unresolvable hostnames are rejected regardless of `p2p.allow_private_urls`.**
- **Discovery intake** (`p2p_client`): announcements validated behind a success-only 5-min TTL cache (2s DNS bound, injectable lookup for tests); rejects metered `invalid`/`blocked` + WARN with peer ID. `asset:8090` never reaches the registry.
- **Refresh-time filter** (`app.endpointSource`): `source=discovered` registry rows re-validated on every listing — **neutralizes the already-poisoned registry row with no data migration**. Operator-configured `datahub_urls` are exempt (may legitimately be cluster-internal). New metric `arcade_teranode_endpoint_refresh_rejected_total{outcome}`; WARN once per URL per process, DEBUG thereafter.
- **Probe ALL endpoints** (`teranode.Client`): the probe loop now probes healthy endpoints too, via a dedicated `consecutiveProbeFailures` counter (trips at `failure_threshold`, reason="probe"). Crucial subtlety: probe success on a *healthy* endpoint resets **only** the probe counter — never the broadcast counters — otherwise a 30s probe cadence against an endpoint whose `/health` answers (but whose `POST /txs` always fails) would perpetually reset the slow-track breaker. Recovery of unhealthy endpoints is byte-for-byte the old contract (any HTTP response counts, incl. 4xx).

## Outcome

- `asset:8090` disappears from `/health` and every broadcast set after rollout.
- A dead-but-resolvable endpoint flips to `healthy: false` in **every** pod within ~90s (3 × 30s probes), no broadcast traffic needed.
- Probe load: ~8 endpoints × ~6 pods / 30s ≈ 1.6 req/s fleet-wide.

## Tests

- `ssrfguard`: ported table + CGNAT boundary cases.
- `p2p_client`: flipped the line-52 "privately-resolving DNS name treated as public" case; headline `asset:8090` unresolvable-rejection; cache = 1 lookup per URL per TTL; upsert-per-announcement preserved.
- `app`: configured-bypass (lookup stub fails test if called), poisoned-row filtering, cache, rejection-never-cached, allow-private opt-in, discovery-off.
- `teranode`: probe demotes an unreachable healthy endpoint; probe successes do NOT reset either broadcast counter (the trap tests); probe-counter consecutive semantics; existing recovery tests pass unchanged.
- `api_server`: end-to-end — `/health` flips a dead endpoint to `healthy:false` driven purely by the probe loop (the exact production complaint).
- Full suite + `magex lint` green. openspec change `harden-datahub-discovery` validates.

## Rollout notes

- No new config keys; no defaults change. `p2p.allow_private_urls` (default false) now also gates DNS-*resolved* private addresses. Static `datahub_urls` are never validated, so local/standalone setups keep working.
- `/health` will start reporting `healthy: false` for genuinely dead endpoints in all pods, and `arcade_teranode_endpoint_healthy` now updates in non-broadcasting pods — check dashboards/alerts keyed on those; `min_healthy_endpoints` WARNs may fire more (signal becoming truthful).
- Follow-ups (documented in the openspec proposal): registry LastSeen pruning; dial-time `ssrfguard.CheckDialAddress` on the broadcast transport to close the DNS-rebinding window.

### CODE REVIEW

File: ssrfguard/ssrfguard.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+//
+//	http://169.254.169.254/latest/meta-data/iam/security-credentials/
+//	http://asset:8090/api/v1  (its own cluster-internal service name)
+//
+// and either turn arcade's propagation path into an SSRF primitive against
+// cloud metadata endpoints / RFC1918 neighbors, or pollute the shared
+// endpoint registry with URLs that can never be reached from this cluster.
+// We block by destination IP rather than hostname so an announcer cannot
+// bypass via DNS, IP literal, or rebinding, and we treat DNS-resolution
+// failure as invalid — an unresolvable name must not enter the registry.
```
[the remark this answers is not in the document]
Comment (galt-tr) [reply ssrfguard/ssrfguard.go]: Fixed in 6d181fb — reworded the threat-model doc: validation-time resolution explicitly does NOT close the rebinding window; that requires wiring CheckDialAddress into the dialer (documented as accepted residual risk / follow-up).

File: metrics/metrics.go [hunk 2]
```diff
@@ -579,7 +589,7 @@ var P2PNodeStatusMessagesTotal = promauto.NewCounter(prometheus.CounterOpts{
 var P2PEndpointDiscoveryTotal = promauto.NewCounterVec(prometheus.CounterOpts{
 	Name: "arcade_p2p_endpoint_discovery_total",
 	Help: "Datahub URL discovery outcomes from peer announcements.",
-}, []string{labelOutcome}) // registered, invalid, no_url, error
+}, []string{labelOutcome}) // registered, invalid, blocked, no_url, error
```
[the remark this answers is not in the document]
Comment (galt-tr) [reply metrics/metrics.go]: Fixed in 6d181fb — added no_store to the documented label set.

File: services/p2p_client/client.go [hunk 3]
```diff
@@ -76,6 +97,14 @@ type Client struct {
...
 	wg            sync.WaitGroup
 	stopOnce      sync.Once
+
+	// lookupIP resolves a hostname during URL validation. Overridable in
+	// tests so validation never hits real DNS; nil is never stored — New
+	// installs the bounded default.
+	lookupIP func(ctx context.Context, host string) ([]net.IP, error)
+	// validatedURLs caches successful validations (raw URL → validatedURL)
+	// so repeat announcements of the same URL skip the DNS round-trip.
+	validatedURLs sync.Map
```
[the remark this answers is not in the document]
Comment (galt-tr) [reply services/p2p_client/client.go]: Fixed in 6d181fb — replaced the sync.Map with a bounded TTL cache (ssrfguard.SuccessCache, cap 512): at capacity it sweeps expired entries and otherwise skips caching, so validation still runs and memory stays bounded under unique-URL spray.

File: app/app.go [hunk 4]
```diff
@@ -476,30 +481,138 @@ func (a chaintracksHeaderReader) GetHeaderByHash(ctx context.Context, hash *chai
...
+	// lookupIP is the DNS seam; nil is never stored — newEndpointSource
+	// installs the bounded default.
+	lookupIP func(ctx context.Context, host string) ([]net.IP, error)
+	// validated caches URLs that passed validation (URL → expiry UnixNano).
+	validated sync.Map
+	// warned tracks URLs already logged at WARN so each rejected row logs
+	// loudly once per process and quietly (DEBUG) on the ~30s re-checks.
+	warnedMu sync.Mutex
+	warned   map[string]struct{}
+	logger   *zap.Logger
```
[the remark this answers is not in the document]
Comment (galt-tr) [reply app/app.go:542]: Fixed in 6d181fb — validated now uses the same bounded SuccessCache, and warned is pruned against every listing, so it's bounded by the registry's current row count rather than everything ever seen.
